0000-experiment-dataframe.md

```diff
@@ -0,0 +1,320 @@
+In this version, `ExpeirmentData.analysis_results()` method still returns a list of `AnalysisResult` by default, and calling with the option `dataframe=True` returns the data frame.
+When data frame is disabled, it internally converts data frame into a list of `AnalysisResult` to return.
+
+`ExperimentData.add_analysis_results()` dispatches the add logic.
+The method signature is updated to `results: Optional[Union[AnalysisResult, List[AnalysisResult]]] = None, **kwargs: Any`.
+If results argument is not None, this must be the legacy analysis result data.
+This implies existing analysis classes can still return this legacy object.
+Legacy result objects are internally converted into a single row of the data frame and added.
+If there are only keyword arguments, we can directly populate the analysis result data frame and bypasses generation of redundant `AnalysisResult` which no longer exists in the `ExperimentData`.
+Existing analysis classes in the Qiskit Experiments namespace will switch to this kwargs style, namely, it must directly populate analysis data from `_run_analysis` method.
+
+`ExperimentData.artifacts()` and `.add_artifacts()` methods are newly implemented with new protected member `ExpeirmentData._artifacts` which is a thread safe list of artifacts.
+This method may provide a filtering function based on the artifact metadata.
+
+_Deprecations_
+- `CurveData` will be replaced with the data frame and thus immediately deprecated. This is in internal use for curve analysis, thus the impact is minimum.
+- Curve analysis options `return_data_points` (curve data) and `return_fit_parameters` (fit status). Now these are moved to artifacts rather than in the analysis results.
+- Add deprecation warning when `ExpeirmentData.analysis_results()` is called with `dataframe=False` (default).
+
+_Remove_
+- None
+
+##### qiskit_experiments v0.7
+In this version we switch the default option of `ExpeirmentData.analysis_results()` to `dataframe=None`. It returns data frame by defaults.
+We must write analysis classes so that it adds analysis results with keyword arguments.
+
+_Deprecations_
+- `AnalysisResult` and `AnalysisResultData`.
+- Deprecate calling `ExperimentData.add_analysis_results()` with results argument.
+- Calling with `ExpeirmentData.analysis_results()` with dataframe argument.
+
+_Remove_
+- `CurveData`
+- Remove fit status and curve data entry generation in analysis results. Curve analysis will complete migration to the artifact at this point.
+
+##### qiskit_experiments v0.8
+In this version we complete migration to data frame and legacy QE classes and methods are completely removed.
+
+_Deprecations_
+- None
+
+_Remove_
+- `AnalysisResult` and `AnalysisResultData`.
+- results argument from the `ExpeirmentData.add_analysis_results()`.
+- dataframe argument from the `ExpeirmentData.analysis_results()`.
+
+
+## Alternative Approaches
+N/A
+
+## Questions
+- What data frame columns must be shown to end users.
```

Comment:
Maybe we can use custom pandas accessors either in service or qiskit experiments for deciding which columns to show.